src/MerchantCenter/MerchantStatuses.php

```diff
@@ -572,32 +583,86 @@ protected function refresh_presync_product_issues(): void {
 		/** @var ProductHelper $product_helper */
-		$product_helper = $this->container->get( ProductHelper::class );
+		$product_helper      = $this->container->get( ProductHelper::class );
+		$visibility_meta_key = $this->prefix_meta_key( ProductMetaHandler::KEY_VISIBILITY );
 
-		foreach ( $validated_mc_statuses as $product ) {
-			$wc_product_id = $product_helper->get_wc_product_id( $product->getProductId() );
-			$status        = $this->get_product_shopping_status( $product );
-			if ( is_null( $status ) ) {
+		foreach ( $statuses as $product_status ) {
+
+			$wc_product_id     = $product_status['product_id'];
+			$mc_product_status = $product_status['status'];
+
+			// Skip if the product does not exist or if the product previously found/validated.
+			if ( ! $wc_product_id || ! empty( $this->product_data_lookup[ $wc_product_id ] ) ) {
+				continue;
+			}
+
+			if ( $this->product_is_expiring( $product_status['expiration_date'] ) ) {
+				$mc_product_status = MCStatus::EXPIRING;
+			}
+
+			$wc_product = $product_helper->get_wc_product_by_wp_post( $wc_product_id );
+			if ( ! $wc_product || 'product' !== substr( $wc_product->post_type, 0, 7 ) ) {
+				// Should never reach here since the products IDS are retrieved from postmeta.
+				do_action(
+					'woocommerce_gla_debug_message',
+					sprintf( 'Merchant Center product %s not found in this WooCommerce store.', $wc_product_id ),
+					__METHOD__,
+				);
 				continue;
 			}
+
+			$this->product_data_lookup[ $wc_product_id ] = [
+				'name'       => get_the_title( $wc_product ),
+				'visibility' => get_post_meta( $wc_product_id, $visibility_meta_key ),
+				'parent_id'  => $wc_product->post_parent,
+			];
+
 			// Products is used later for global product status statistics.
-			$this->product_statuses['products'][ $wc_product_id ][ $status ] = 1 + ( $this->product_statuses['products'][ $wc_product_id ][ $status ] ?? 0 );
+			$this->product_statuses['products'][ $wc_product_id ][ $mc_product_status ] = 1 + ( $this->product_statuses['products'][ $wc_product_id ][ $mc_product_status ] ?? 0 );
 
 			// Aggregate parent statuses for mc_status postmeta.
-			$wc_parent_id = intval( $this->product_data_lookup[ $wc_product_id ]['parent_id'] );
+			$wc_parent_id = $this->product_data_lookup[ $wc_product_id ]['parent_id'];
 			if ( ! $wc_parent_id ) {
 				continue;
 			}
-			$this->product_statuses['parents'][ $wc_parent_id ][ $status ] = 1 + ( $this->product_statuses['parents'][ $wc_parent_id ][ $status ] ?? 0 );
+			$this->product_statuses['parents'][ $wc_parent_id ][ $mc_product_status ] = 1 + ( $this->product_statuses['parents'][ $wc_parent_id ][ $mc_product_status ] ?? 0 );
+
 		}
 	}
 
+	/**
+	 * Update the product status statistics.
+	 */
+	public function update_product_stats() {
+		$this->mc_statuses = [];
+		// Update each product's mc_status and then update the global statistics.
+		$this->update_product_mc_statuses();
+		$this->update_mc_statuses();
```

Comment:
The naming of these two functions makes me have to look up what each one does. Since they are just internal functions would a more descriptive name help here?

```suggestion
		$this->update_products_meta_with_mc_status();
		$this->update_mc_status_statistics();
```